Lib/test/test_zipapp.py

```diff
@@ -417,6 +485,63 @@ def test_info_error(self):
         # Program should exit with a non-zero return code.
         self.assertTrue(cm.exception.code)
 
+    def test_cmdline_create_with_include_pattern(self):
+        source = self.tmpdir / 'source'
+        source.mkdir()
+        (source / '.DS_Store').touch()
+        (source / 'zed.py').touch()
+        (source / 'bin').mkdir()
+        (source / 'bin' / 'qux').touch()
+        (source / 'bin' / 'baz').touch()
+        (source / '__main__.py').touch()
+
+        args = [str(source), '--include-pattern', r'.*\.py']
+        zipapp.main(args)
+        target = source.with_suffix('.pyz')
+        self.assertTrue(target.is_file())
+
+        with zipfile.ZipFile(target, 'r') as zf:
+            self.assertEqual(zf.namelist(),
+                    ["__main__.py", "zed.py"])
+
+    def test_cmdline_create_with_exclude_pattern(self):
+        source = self.tmpdir / 'source'
+        source.mkdir()
+        (source / '.DS_Store').touch()
+        (source / 'zed.py').touch()
+        (source / 'bin').mkdir()
+        (source / 'bin' / 'qux').touch()
+        (source / 'bin' / 'baz').touch()
+        (source / '__main__.py').touch()
+
+        args = [str(source), '--exclude-pattern', r'.*\.py']
+        zipapp.main(args)
+        target = source.with_suffix('.pyz')
+        self.assertTrue(target.is_file())
+
+        with zipfile.ZipFile(target, 'r') as zf:
+            self.assertEqual(zf.namelist(),
+                [".DS_Store", "bin/", "bin/baz", "bin/qux"])
+
+    def test_cmdline_create_with_include_and_exclude_pattern(self):
+        source = self.tmpdir / 'source'
+        source.mkdir()
+        (source / '.DS_Store').touch()
+        (source / 'zed.py').touch()
+        (source / 'bin').mkdir()
+        (source / 'bin' / 'qux').touch()
+        (source / 'bin' / 'baz').touch()
+        (source / '__main__.py').touch()
+
+        args = [str(source), '--include-pattern', r'.*\.py', '--exclude-pattern', r'.*z.*']
```

Comment:
Can `self.tmpdir` contain a `z`? If so, I believe this exclude pattern would ignore all files (since AFAICS it's matched against absolute paths, not against paths relative to the `source` directory)

Reply:
Thanks @godlygeek , note that this is fixed.